beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -2841,17 +2846,16 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         // If the write fails, revert fork choice to the version from disk, else we can
         // end up with blocks in fork choice that are missing from disk.
         // See https://github.com/sigp/lighthouse/issues/2028
-        let mut ops: Vec<_> = confirmed_state_roots
-            .into_iter()
-            .map(StoreOp::DeleteStateTemporaryFlag)
-            .collect();
+        ops.extend(
+            confirmed_state_roots
+                .into_iter()
+                .map(StoreOp::DeleteStateTemporaryFlag),
+        );
         ops.push(StoreOp::PutBlock(block_root, signed_block.clone()));
         ops.push(StoreOp::PutState(block.state_root(), &state));
         let txn_lock = self.store.hot_db.begin_rw_transaction();
```

Comment:
I'm pretty sure this introduces a deadlock because `do_atomically` doesn't execute in a spawned thread.

Reply:
Could you elaborate a little?

As far as I can tell there's no deadlock because `.begin_rw_transaction` locks its own mutex, and the only other place this mutex is locked is here:

https://github.com/sigp/lighthouse/blob/38514c07f222ff7783834c48cf5c0a6ee7f346d0/beacon_node/beacon_chain/src/block_verification.rs#L1274

In both cases we lock `begin_rw_transaction`'s mutex first and then call `do_atomically` which locks the block cache, which means our lock ordering is always:

```
lock(A); // begin_rw_transaction
lock(B); // block_cache mutex
unlock(B);
unlock(A);
```

So there's no risk of deadlock from lock ordering (i.e. no instances of `lock(B); lock(A)`).

Regarding threading, I think you might be alluding to the deadlocks that can arise from mutexes held across await? I.e. the reason for "async" mutexes, described here: https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html. That also isn't an issue for us because `import_block` isn't an async function and we don't hold the mutex across any `await` points. We run `import_block` on the Tokio thread pool for blocking tasks, which AFAIU behave like regular OS threads (they have their context restored when preempted).

https://github.com/sigp/lighthouse/blob/38514c07f222ff7783834c48cf5c0a6ee7f346d0/beacon_node/beacon_chain/src/beacon_chain.rs#L2655-L2672

Reply:
Ah, yes! I was think of the scopes wrong for "scoped lock" (parking_lot docs). Nothing is actually being called on `tx`, so it doesn't have to be in the same scope just the same thread, got it, which the nested scope of `do_atomically` is. My mention on threading was simply based on the comment to the method
```rust
    /// Return a mutex guard that can be used to synchronize sensitive transactions.
    ///
    /// This doesn't prevent other threads writing to the DB unless they also use
    /// this method. In future we may implement a safer mandatory locking scheme.
    fn begin_rw_transaction(&self) -> MutexGuard<()>;
```
Thank you for the explanation.